test/integration/provider/rti.go

```diff
@@ -0,0 +1,149 @@
+	v1 "k8s.io/api/core/v1"
+)
+
+var (
+	// cluster tag and cluster values are specifically used for integration test
+	// in case the test is run against non seed cluster then the supplied MachineClass
+	// is expected to have these tags set so that the machines from this suite won't be
+	// orphan collected.
+	clusterTag      = "kubernetes.io-role-integration-test"
+	clusterTagValue = "1"
+)
+
+// ResourcesTrackerImpl implements the Resource Tracker Interface from the Integration test suite
+type ResourcesTrackerImpl struct {
+	ClusterName      string
+	InitialVolumes   []string
+	InitialInstances []string
+	InitialMachines  []string
+	MachineClass     *v1alpha1.MachineClass
+	ResourceGroup    string
+	SecretData       map[string][]byte
+}
+
+// InitializeResourceTracker is the constructor of ResourceTrackerImpl
+func (r *ResourcesTrackerImpl) InitializeResourcesTracker(machineClass *v1alpha1.MachineClass, secretData map[string][]byte, clusterName string) error {
+
+	r.MachineClass = machineClass
+	r.SecretData = secretData
+	r.ClusterName = clusterName
+
+	providerSpec, err := provider.DecodeProviderSpecAndSecret(machineClass, &v1.Secret{Data: secretData})
+	if err != nil {
+		return err
+	}
+	r.ResourceGroup = providerSpec.ResourceGroup
+
+	clients, err := getAzureClients(secretData)
+	if err != nil {
+		return err
+	}
+
+	instances, err := getVMsWithTag(clients, clusterTag, clusterTagValue, machineClass, r.ResourceGroup, secretData)
+	if err == nil {
+		r.InitialInstances = instances
+		volumes, err := getAvailableDisks(clients, clusterTag, clusterTagValue, r.ResourceGroup)
+		if err == nil {
+			r.InitialVolumes = volumes
+			r.InitialMachines, err = getMachines(machineClass, secretData)
+			return err
+		}
+		return err
+	}
+	return err
+}
+
+// IsOrphanedResourcesAvailable checks whether there are any orphaned resources left.
+// If yes, then prints them and returns true. If not, then returns false
+func (r *ResourcesTrackerImpl) IsOrphanedResourcesAvailable() bool {
+	afterTestExecutionInstances, afterTestExecutionAvailVols, afterTestExecutionAvailmachines, err := r.probeResources()
+	//Check there is no error occured
+	if err == nil {
+		orphanedResourceInstances := differenceOrphanedResources(r.InitialInstances, afterTestExecutionInstances)
+		if orphanedResourceInstances != nil {
+			fmt.Println("orphaned instances were:", orphanedResourceInstances)
+			return true
+		}
+		orphanedResourceAvailVols := differenceOrphanedResources(r.InitialVolumes, afterTestExecutionAvailVols)
+		if orphanedResourceAvailVols != nil {
+			fmt.Println("orphaned volumes were:", orphanedResourceAvailVols)
+			return true
+		}
+		orphanedResourceAvailMachines := differenceOrphanedResources(r.InitialMachines, afterTestExecutionAvailmachines)
+		if orphanedResourceAvailMachines != nil {
+			fmt.Println("orphaned volumes were:", orphanedResourceAvailMachines)
+			return true
+		}
```

Comment:
We should not `return true` in if statements
scenario:
If before tests there was an orphan VM lets say `VM1` we would note it down in r.InitialInstances and would delete it, and so it won't be there when we run probeResources() after the test. 
Now `differenceOrphanedResources()` will conclue that an orphan resource is still available(because it calculates the difference b/w orphan instances before and after test), and we'll print the line `orphaned instances were`..... But if we return true we won't be able to execute the other ifs
Thats what I have done, I have removed the `return true` statements and it works for me in the scenario when we did delete an orphan resource.

Reply:
Yes, good catch. I will remove them

Reply:
Even though your approach looked good to me yesterday. I was not fully convinced with it because, that way, `CleanUp` wouldn't do its job right. I mean, if there are orphaned resources left then `CleanUp` test case should fail and not just print the resources. To handle the case you explained, we need to change the way the get methods are implemented today.

Here is an approach I put forward. I tried and it works too!

Generic flow would look like this

- Get Resources from Resource Group
  - Filter Resources associated with Integration Test Tags
  - Delete the Resources
  - If Deletion is not successful
    - add to the list of Resources (which will be then stored in initial Resources)

Reply:
Good points. I understand your concern in point 2, but I would recommend that the test case of cleanup() should fail instead of just printing so, we can have some logic like

```
if afterTestExecutionInstances != nil || afterTestExecutionAvailVols != nil || afterTestExecutionAvailmachines != nil {
    // print all three anyway. If one or more of them is empty then they print nothing
    return true // so that the cleanup fails and repeats until timeout
}
```

And we can remove the `differenceOrphanedResources()` totally. Because, anyway, the probeResources would only return the actual orphaned resources.